recipes/DebuggingLog/src/Main.purs

```diff
@@ -0,0 +1,129 @@
+import Data.Int (toNumber)
+import Data.Tuple (Tuple(..))
+import Debug.Trace (spy, traceM)
+import Effect (Effect)
+import Effect.Aff (Milliseconds(..), delay, launchAff_)
+import Effect.Class (liftEffect)
+import Effect.Console (log)
+import Effect.Random (randomInt)
+
+data MyADT = MyADT Int (Tuple Int (Array String)) { foo :: String }
+
+main :: Effect Unit
+main = do
+  log "When we are in the 'Effect' monad, we can print content to the console."
+  launchAff_ do
+    liftEffect $ log $ "We can still print values to the console as long as \
+                       \ the monad in question implements the MonadEffect \
+                       \type class. Since `Aff` implements MonadEffect, we \
+                       \ can lift that effect into `Aff`."
+
+  log "However, there are times when we want to debug some code and wish \
+      \to use print-based debugging. Since PureScript is pure, how do we \
+      \do that?"
+
+  usingSpy
+
+  usingTraceMInEffect
+  usingTraceMInST
+  usingTraceMInAff
+  usingTraceMInAffRace
+
+usingSpy :: Effect Unit
+usingSpy = do
+  log "usingSpy"
+  -- `spy` returns the value it receives. However, it prints that value
+  -- to the console before it "returns."
+  --    spy :: forall a. DebugWarning => String -> a -> a
+  -- You can use this in a `let` clause pretty reliably
+  let
+    x = 5
+    y = spy "y" 8
+    adt = spy "adt" $ MyADT 1 (Tuple 4 ["a", "b"]) { foo: "foo value" }
+    function = spy "function" $ \intValue -> show $ 4 + intValue
+
+    -- quick way of debugging something without showing it or using a
+    -- variable name
+    _ = spy "debug for me what x + y is" $ x + y
+    arrayOfStrings = spy "debug some array" $ map show [1, 2, 3, 4, 5]
+
+    -- quickly drop-in `spy` to see every step of a recursive function
+    fact :: Int -> Int -> Int
+    fact 0 acc = acc
+    fact n acc = fact (spy "n" (n - 1)) (spy "acc" (acc * n))
+
+    _ = fact 5 1
+
+  launchAff_ do
+    let
+      y = spy "y in Aff" 8
+      adt = spy "adt in Aff" $ MyADT 1 (Tuple 4 ["a", "b"]) { foo: "foo value" }
+      function = spy "function in Aff" $ \intValue -> show $ 4 + intValue
+    pure unit
+
+  log $ "Thus, spying is an effective way of quickly adding debugging where \
+        \you need it without affecting any other part of your code. \
+        \Note: you should not do this in production code. Use a proper logger \
+        \in production code."
+
+usingTraceMInEffect :: Effect Unit
+usingTraceMInEffect = do
+  log "usingTraceMInEffect"
+  log "But what if you don't want to depend on `console` to do temporary \
+      \printing-based debugging? We can use traceM instead."
+  traceM "logging to the console without using `log`"
+
+usingTraceMInST :: Effect Unit
+usingTraceMInST = do
+  let localMutationResult = ST.run do
+        localReference <- ST.new 0
+        traceM localReference
+        four <- ST.modify (_ + 4) localReference
+        traceM four
+        ST.write (four + 2) localReference
```

Comment:
I'm not seeing the advantage of `traceM` over `spy` in this case (assuming there's a "tagless" version of `spy` available).
```purs
usingTraceMInST = do
  let localMutationResult = ST.run do
        localReference <- ST.new 0
        four <- ST.modify (_ + 4) $ spy localReference
        ST.write ((spy four) + 2) localReference
```

Maybe the benefit is convenience in controling the order that values are printed in and being able to easily add/remove these debug prints.

Reply:
Honestly, I'm not sure why `traceM` exists when we have `spy`. Maybe because of the following:
```
let _ = spy "foo" foo
-- vs
traceM foo
```